src/OSWindow-SDL2/OSSDLPasteEventFilter.class.st

```diff
@@ -0,0 +1,155 @@
+Class {
+	#name : #OSSDLPasteEventFilter,
+	#superclass : #Object,
+	#instVars : [
+		'next',
+		'lastEvents',
+		'currentState'
+	],
+	#pools : [
+		'SDL2Constants'
+	],
+	#category : #'OSWindow-SDL2-Base'
+}
+
+{ #category : #'last events' }
+OSSDLPasteEventFilter >> addEvent: aEvent [
+
+	self lastEvents add: aEvent
+]
+
+{ #category : #'events-processing' }
+OSSDLPasteEventFilter >> changeEventsToIncludeModifier [
+
+	| modifierToUse defaultModifier |
+
+	"We modify the KeyUp/KeyDown events to include the correct modifier.
+	This is because tools to handle copy/paste use virtual events, and SDL is dropping the modifier.
+	
+	The correct modifier is calculated depending of the current platform modifier"
+
+	modifierToUse := 0.
+	defaultModifier := OSPlatform current defaultModifier.
+	
+	defaultModifier = KMModifier command ifTrue: [ modifierToUse := KMOD_GUI ].
+	defaultModifier = KMModifier alt ifTrue: [ 	modifierToUse := KMOD_LALT ].
+	defaultModifier = KMModifier ctrl ifTrue: [ modifierToUse := KMOD_LCTRL ].
+
+	self lastEvents do: [ :anEvent | 
+		(anEvent isKeyUpEvent or: [ anEvent isKeyDownEvent ]) 
+			ifTrue: [ anEvent keysym mod: (anEvent keysym mod bitOr: modifierToUse) ] ]
+]
+
+{ #category : #accessing }
+OSSDLPasteEventFilter >> currentState [
+
+	^ currentState ifNil: [ currentState := 1 ]
+]
+
+{ #category : #accessing }
+OSSDLPasteEventFilter >> currentState: aValue [
+
+	currentState := aValue
+]
+
+{ #category : #'events-processing' }
+OSSDLPasteEventFilter >> dispatchEvent: aSDLEvent [
+	
+	"This filter will handle the case when an external Clipboard handling tool is used.
```

Comment:
I don't know if it is worth to note that this seems to be a OSX SDL2 bug.

Reply:
Yes, I will add that